source/common/init/watcher_impl.h

```diff
@@ -24,23 +25,35 @@ using ReadyFn = std::function<void()>;
 class WatcherHandleImpl : public WatcherHandle, Logger::Loggable<Logger::Id::init> {
 private:
   friend class WatcherImpl;
+
+  // Ctor with std::function<void()> callback function.
   WatcherHandleImpl(absl::string_view handle_name, absl::string_view name,
                     std::weak_ptr<ReadyFn> fn);
 
+  // Ctor with std::function<void(absl::string_view)> callback function.
+  WatcherHandleImpl(absl::string_view handle_name, absl::string_view name,
+                    std::weak_ptr<ReadyFnSendName> fn);
+
 public:
-  // Init::WatcherHandle
+  // Init::WatcherHandle.
   bool ready() const override;
 
 private:
   // Name of the handle (either the name of the target calling the manager, or the name of the
-  // manager calling the client)
+  // manager calling the client).
   const std::string handle_name_;
 
-  // Name of the watcher (either the name of the manager, or the name of the client)
+  // Name of the watcher (either the name of the manager, or the name of the client).
   const std::string name_;
 
-  // The watcher's callback function, only called if the weak pointer can be "locked"
+  // When the function constructed is ReadyFnSendName, set this to true.
+  bool fn_send_name_;
+
+  // The watcher's callback function, only called if the weak pointer can be "locked".
   const std::weak_ptr<ReadyFn> fn_;
+
+  // The watcher's callback function, only called if the weak pointer can be "locked".
+  const std::weak_ptr<ReadyFnSendName> fn_sn_;
```

Comment:
Not sure I agree with this design. It basically says "if you want this watcher to send a name, construct me with a `ReadyFnSendName`, and if not, ehhh whatever." That means you'll have inconsistent usage across the codebase. I think you want your design to enforce consistency so that the information is always there when you need it.

Reply:
Thanks for the advice. Here the two options of ctors of the watcherImpl are mainly designed to support the old codes and my new codes. 

But the main idea is that the manager could get the information of target names. As in my current code, the manager is constructed with a watcher(with [](string_view){} as [function](https://github.com/ASOPVII/envoy/blob/init-manager-query-unready-targets/source/common/init/manager_impl.cc#L11-L15)), and it is when the target calls its initialize that the watcher gets a name, (code [link](https://github.com/ASOPVII/envoy/blob/init-manager-query-unready-targets/source/common/init/target_impl.cc#L16)). Therefore, even the old code does not pass a parameter, it could be regarded as an empty string_view, which does not influence the old logic. In addition, if people want to use this feature, they could try to add string_view parameter to the ctor of watcher, so that the manager could get real target names. I think this

Reply:
The best way to understand this pr is that it not only maintains the old version of watcher ctor, but also supports new ctor, which provides the manager a method to know target names. Even though that in the old code base, watchers are most init with ```[](){}```, it does only have effect on the old tests. Since now the manager has been changed to receive a new type of function, the manager will always get the information of target names. I think the code is not only consistent with the old version, but helps to provide the features. Please notice that managerImpl ctor is the main point. WatcherImpl ctors only help to support both versions(most old test case where watchers are init with ```[](){}```.)

Reply:
> Not sure I agree with this design. It basically says "if you want this watcher to send a name, construct me with a `ReadyFnSendName`, and if not, ehhh whatever." That means you'll have inconsistent usage across the codebase. I think you want your design to enforce consistency so that the information is always there when you need it.

Thanks for the review. The point is not focused on WatcherImpl. Please look at the impl of ManagerImpl, now the watcher_ of the ManagerImpl is constructed with ```[](string_view){}```, so it should always get the target names. I believe there is no inconsistency, even though the watchers in old tests are mostly constructed with ```[](){}```, it does not influence the managerImpl.

Reply:
I don't like this either. I guess you would be able to roll `ReadyFnSendName` into `ReadyFn` by capturing the name in the lambda, and the rest of WatcherHandle changes won't be necessary.